src/Microsoft.ML.Data/Transforms/KeyToVectorTransform.cs

```diff
@@ -69,134 +75,109 @@ public bool TryUnparse(StringBuilder sb)
         }
 
-        private KeyToVectorTransform(IHost host, ModelLoadContext ctx, IDataView input)
-            : base(host, ctx, input, TestIsKey)
+        private string TestIsKey(ColumnType type)
         {
-            Host.AssertValue(ctx);
+            if (type.ItemType.KeyCount > 0)
+                return null;
+            return "Expected Key type of known cardinality";
+        }
 
-            // *** Binary format ***
-            // <prefix handled in static Create method>
-            // <base>
-            // for each added column
-            //   byte: bag as 0/1
-            Host.AssertNonEmpty(Infos);
-            int size = Infos.Length;
-            _bag = new bool[size];
-            _concat = new bool[Infos.Length];
-            _types = new VectorType[size];
-            for (int i = 0; i < size; i++)
+        public KeyToVectorTransform(IHostEnvironment env, IDataView input, ColumnInfo[] columns) :
+            base(Contracts.CheckRef(env, nameof(env)).Register(RegistrationName), GetColumnPairs(columns))
+        {
+            var infos = CreateInfos(input.Schema);
+            _bags = new bool[infos.Length];
+            _valueCounts = new int[infos.Length];
+            _sizes = new int[infos.Length];
+
+            for (int i = 0; i < infos.Length; i++)
             {
-                _bag[i] = ctx.Reader.ReadBoolByte();
-                ComputeType(this, Source.Schema, i, Infos[i], _bag[i], Metadata,
-                    out _types[i], out _concat[i]);
+                _bags[i] = columns[i].Bag;
+                _sizes[i] = infos[i].TypeSrc.ItemType.KeyCount;
+                _valueCounts[i] = infos[i].TypeSrc.ValueCount;
             }
-            Metadata.Seal();
         }
+        public const string LoaderSignature = "KeyToVectorTransform";
+        public const string UserName = "KeyToVectorTransform";
+        internal const string Summary = "Converts a key column to an indicator vector.";
 
-        public static KeyToVectorTransform Create(IHostEnvironment env, ModelLoadContext ctx, IDataView input)
+        private static VersionInfo GetVersionInfo()
         {
-            Contracts.CheckValue(env, nameof(env));
-            var h = env.Register(RegistrationName);
-            h.CheckValue(ctx, nameof(ctx));
-            h.CheckValue(input, nameof(input));
-            return h.Apply("Loading Model",
-                ch =>
-                {
-                    // *** Binary format ***
-                    // int: sizeof(Float)
-                    // <remainder handled in ctors>
-                    int cbFloat = ctx.Reader.ReadInt32();
-                    ch.CheckDecode(cbFloat == sizeof(Float));
-                    return new KeyToVectorTransform(h, ctx, input);
-                });
+            return new VersionInfo(
+                modelSignature: "KEY2VECT",
+                //verWrittenCur: 0x00010001, // Initial
+                verWrittenCur: 0x00010002, // Convert to Estimators
```

Comment:
>verWrittenCur: 0x00010002, // Convert to Estimators [](start = 16, length = 51)

As far as I know it has not been necessary to break backwards compatibility so far, why is it necessary here?

Reply:
No more backward incompatible.

---
In reply to: [216140582](https://github.com/dotnet/machinelearning/pull/858#discussion_r216140582) [](ancestors = 216140582)